### PR TITLE
add support for riscv

### PR DESCRIPTION
**Pull request by yuzibo**

### CODE REVIEW

**Comment by yuzibo** [comment]
I can built it on my real riscv64 hardware(Unmatched board)

**Comment by munroesj52** [comment]
Apologies for the delay. I have been fighting cold. 

So you are adding specific config for **__riscv_xlen == 64** where previously all riscv was assummed to be 64-bit?

This implies that there are **__riscv_xlen == 32** machines which fall into the default config parms.

Have your tested the 32-bit case to insure that the default config is OK?

I have no way to test this myself.

**Comment by yuzibo** [comment]
> Apologies for the delay. I have been fighting a cold.

No problem, take care.
> 
> So you are adding specific config for **__riscv_xlen == 64** where previously all riscv was assummed to be 64-bit?
> 
> This implies that there are **__riscv_xlen == 32** machines which fall into the default config parms.
> 
> Have your tested the 32-bit case to insure that the default config is OK?

Ah, sorry, I do not have any riscv32 hardware until now from my side. So this time I only support riscv64 explicitly. I look back the code, it may be easy to support riscv32 if drop `__riscv_xlen == 64` constraint. 

So could we support riscv64 first and then to add riscv32 support if there is riscv32 hardware available? yeah, this is not friendly for maintainer.:)

 > 
> I have no way to test this myself.

**Comment by munroesj52** [comment]
Sorry for the delay again, still recovering.

There are three issues that are target dependent.

1) Endian:  This usually works if the compiler defines __ORDER_LITTLE_ENDIAN__,  __ORDER_BIG_ENDIAN__, and __BYTE_ORDER__. But there was a time when that was not always guaranteed. 

So does riscv support both endians for just little endian?

2) for sasatom.h and variants  We need a interger type with the same size as __UINTPTR_TYPE__. Depending on the compiler (compiler version and target) this might be int, long int, long long int. Usually long unsigned int is the correct choice. but not always (especially for ISA with 32-bit in 64 modes). I will have to look at iso-C and verify when  __UINTPTR_TYPE__ was defined and mandated.

Using __UINTPTR_TYPE__ might help sasatom_generic.h become more generic.  But I am not convinced that will be enough for all platforms. I suspect will still need __WORDSIZE_64/__WORDSIZE_32 for a while.

3) The final issue is how the kernel allocates process storage. This is unique to the hardware platform and the bitness (32/64) and kernal version. SPHDE needs a block of addresses that are not likely to be allocated by the kernel for main/dyn-libs/heap/stack.

As far as I can tell your patch is correct for  __riscv_xlen == 64 but a miss-configurion for __riscv_xlen == 32 as
- __SAS_BASE_ADDRESS is out of range for 32-bit 
-  __WORDSIZE_64  is inappropriate doe 32-bit
- So the #if __riscv_xlen == 64 conditional is required.
- I would like to see a #else under this with appropriate values for  __riscv_xlen == 32 

What is not known if the default sasconf works for __riscv_xlen == 32? Or even what the apprproaite values are. I suspect it has not been attempted.

I am thinking it is better to commit your patch for 64-bit riscv and  hope that the 32-bit riscv community will submit an update.

Do you agrre?

**Comment by munroesj52** [comment]
I am told that [RISC-V Lab](https://rvlab.org/ ) provides open source access to 32-bit RISC-V hardware. If you are interested.

**Comment by yuzibo** [comment]
Sorry for so late reply.

> So does riscv support both endians for just little endian?

If from riscv [spec](https://github.com/riscv-non-isa/riscv-elf-psabi-doc/releases/download/draft-20240325-948463cd5dbebea7c1869e20146b17a2cc8fda2f/riscv-abi.pdf) view, this is pre-keep by big endian's definition, see 8.1 section from above link. but there are notes from `Status of ABI`, `Procedure Calling Convention` :

```
ABI for big-endian is NOT included in this specification, we intend to define that in
future version of this specification
```
Moreover, I worked as a Debian [riscv](https://wiki.debian.org/Ports/riscv64) porter, from linux distro's view, there is only little endian support for riscv64 as far as I know. So I am not sure we how to deal with the situation.

> Using **UINTPTR_TYPE** might help sasatom_generic.h become more generic. But I am not convinced that will be enough for all platforms. I suspect will still need __WORDSIZE_64/__WORDSIZE_32 for a while.

Thanks for sharing about it.  if we consider only support riscv64, is [__WORDSIZE_64](https://github.com/sphde/sphde/pull/70/files#diff-47c2b49fb83417716a4f453e47482d96c91c6955eb076fdaf479ac7b2a9e5d42R121) okay?



> What is not known if the default sasconf works for __riscv_xlen == 32? Or even what the apprproaite values are. I suspect it has not been attempted.

yeah, here is one code example for [this](https://sources.debian.org/src/util-linux/2.39.3-11/include/audit-arch.h/?hl=23#L20)

Fortunately, I get one rv32 testing env from the help of plct lab. Now I am testing it.

**Comment by yuzibo** [comment]
> 3\. The final issue is how the kernel allocates process storage. This is unique to the hardware platform and the bitness (32/64) and kernal version. SPHDE needs a block of addresses that are not likely to be allocated by the kernel for main/dyn-libs/heap/stack.

yeah, I realize your comment finally. 

Based on [f3be0d8](https://github.com/sphde/sphde/commit/f3be0d8d14b939fc607a2db3688016291f408e65), now sphde can built on riscv32 and riscv64.

But unfortunately, now both check [failed](https://gist.github.com/yuzibo/34a83354a90d17674685a4d1aa205cdb). Not sure how to promise to pass these test cases. Maybe need to modify kernel config first?

**Comment by munroesj52** [comment]
What is sizeof (long int) for riscv32/64?

try compiling sassim.ccp with -E -dD to see what the conditional code is actually doing.

**Comment by munroesj52** [comment]
I will be leaving on vacation next week and will be out for basically the month of April.

Tulio can you watch this while I am out.

**Comment by munroesj52** [comment]
> But unfortunately, now both check [failed](https://gist.github.com/yuzibo/34a83354a90d17674685a4d1aa205cdb). Not sure how to promise to pass these test cases. Maybe need to modify kernel config first?` 

No

sasconf.h must provide (Region) values for __SAS_BASE_ADDRESS , RegionSize,  SegmentSize,  __SAS_SHMAP_MAX that valid for the Arch and kernel where the application is running.

__WORDSIZE_64 or __WORDSIZE_32 need to be defined appropriately. If neither is defined there is CPP at the end of sasconf.h that will #define __WORDSIZE_32 along with the default (32-bit) region values. (which you dont want for riscv-64)

**Comment by munroesj52** [comment]
Also which test cases are failing?

**Comment by tuliom** [comment]
> Also which test cases are failing?

In my tests, the following tests are failing on riscv64:

```
FAIL: sassim_t
FAIL: sascompoundheap_t
FAIL: sasseg_t
FAIL: sphlflogger_tt
FAIL: sphlflogger_ttt
FAIL: sphcontext_t
FAIL: sasindex_t
FAIL: sasindex_tt
FAIL: sasstringbtree_t
FAIL: sasstringbtree_tt
FAIL: sphsinglepcqueue_tt
FAIL: sphsinglepcqueue_ttt
FAIL: sphdirectpcqueue_ttt
============================================================================
Testsuite summary for libsphde 1.4.0
============================================================================
# TOTAL: 23
# PASS:  10
# SKIP:  0
# XFAIL: 0
# FAIL:  13
# XPASS: 0
# ERROR: 0
```

All the tests that failed are having issues to allocate memory using `mmap`.

**Comment by munroesj52** [comment]
I am back. Looks like last interaction war April 4.

Any progress?

**Comment by munroesj52** [comment]
I will would like to see the memory map for each platform. Details like where the main executable and libraries load by default. Also the location of the process heap, stack. and VDSO segments.

You get this while running on the platform with:
$ cat /proc/self/maps

File: src/sasconf.h [hunk 1]
```diff
@@ -116,6 +116,21 @@
...
+  # define     __SAS_SHMAP_MAX         0x01000000UL    /*  16MB */
+  #elif __riscv_xlen == 64
+  # define     __WORDSIZE_64
+  # define     __SAS_BASE_ADDRESS      0x4000000000L
+  # define     RegionSize              0x2000000000L   /* 128GB */
+  # define     SegmentSize             0x0010000000L   /* 256MB */
+  # define     __SAS_SHMAP_MAX         0x0001000000L   /* 16MB */
+  #endif
+#endif
+
```
[the remark this answers is not in the document]
Comment (yuzibo) [reply src/sasconf.h:133]: @tuliom  could you give me the clearer value for `sasconf.h` so that I can align with you then push here?

**Comment by yuzibo** [comment]
> I am back. Looks like last interaction war April 4.
:)
> 
> Any progress?

Sorry for the late.

> $ cat /proc/self/maps
```
vimer@debci-02:~/build/sphde/sphde$ cat /proc/self/maps
2ab174a000-2ab1752000 r-xp 00000000 103:03 467                           /usr/bin/cat
2ab1752000-2ab1753000 r--p 00007000 103:03 467                           /usr/bin/cat
2ab1753000-2ab1754000 rw-p 00008000 103:03 467                           /usr/bin/cat
2aba516000-2aba537000 rw-p 00000000 00:00 0                              [heap]
3fb9000000-3fb95ee000 r--p 00000000 103:03 299                           /usr/lib/locale/locale-archive
3fb9659000-3fb967b000 rw-p 00000000 00:00 0
3fb967b000-3fb979d000 r-xp 00000000 103:03 2271                          /usr/lib/riscv64-linux-gnu/libc.so.6
3fb979d000-3fb97a0000 r--p 00122000 103:03 2271                          /usr/lib/riscv64-linux-gnu/libc.so.6
3fb97a0000-3fb97a2000 rw-p 00125000 103:03 2271                          /usr/lib/riscv64-linux-gnu/libc.so.6
3fb97a2000-3fb97b0000 rw-p 00000000 00:00 0
3fb97b8000-3fb97ba000 r--p 00000000 00:00 0                              [vvar]
3fb97ba000-3fb97bb000 r-xp 00000000 00:00 0                              [vdso]
3fb97bb000-3fb97d7000 r-xp 00000000 103:03 1760                          /usr/lib/riscv64-linux-gnu/ld-linux-riscv64-lp64d.so.1
3fb97d7000-3fb97d9000 r--p 0001b000 103:03 1760                          /usr/lib/riscv64-linux-gnu/ld-linux-riscv64-lp64d.so.1
3fb97d9000-3fb97db000 rw-p 0001d000 103:03 1760                          /usr/lib/riscv64-linux-gnu/ld-linux-riscv64-lp64d.so.1
3fde7a3000-3fde7c4000 rw-p 00000000 00:00 0                              [stack]

```

File: src/sasconf.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (yuzibo) [reply src/sasconf.h:133]: I tried it but failed so I think I am wrong something.

**Comment by munroesj52** [comment]
Ok the this maps helps.

So this like 64-bit system and kernel?

The top of stack is at 0x3fde7c4000. Which normally (near the) top of the supported address space.

You placed __SAS_BASE_ADDRESS  at  0x4000000000L. Which is above the stack. As far as I know (the kernels may have changed)  this is not allowed.

So  __SAS_BASE_ADDRESS has to go:
- between the heap (with room to grow the heap) and the dynamic libraries
- Or below the executable /usr/bin/cat  at 2ab174a000 but not at 0x000000000L. (the first page/segment is reserved as zero/no access).
- But
  - The RegionSize must be a power of 2.
  - And the __SAS_BASE_ADDRESS must be a integer multiple of RegionSize.
  - And you need little more room above the "region" for __SAS_SHMAP_MAX (whihc is used for the global lock table.

So the following should work:
+  # define     __SAS_BASE_ADDRESS      0x3000000000L
+  # define     RegionSize              0x0800000000L

OR 
+  # define     __SAS_BASE_ADDRESS      0x1000000000L
+  # define     RegionSize              0x1000000000L

Youl will ned to run
    $ cat /proc/self/maps

On a 32-bit system/kernel to get the values for __riscv_xlen == 32

**Comment by munroesj52** [comment]
Looks like the following would work:
#if __riscv_xlen == 32
# define     __SAS_BASE_ADDRESS      0x80000000UL    /* 2.0GB */
# define     RegionSize              0x10000000UL    /* 256MB */
# define     SegmentSize             0x01000000UL    /*  16MB */
# define     __SAS_SHMAP_MAX         0x01000000UL    /*  16MB */  
#elif __riscv_xlen == 64
...

I would avoid __SAS_BASE_ADDRESS      0x70000000UL because it would not leave much room for the process heap.

or
#if __riscv_xlen == 32
# define     __SAS_BASE_ADDRESS      0x20000000UL    /* 0.5GB */
# define     RegionSize              0x20000000UL    /* 512MB */
# define     SegmentSize             0x01000000UL    /*  16MB */
# define     __SAS_SHMAP_MAX         0x01000000UL    /*  16MB */  
#elif __riscv_xlen == 64
...

The default for 32-bit (__SAS_BASE_ADDRESS      0x60000000UL) might work for a little bit but would eventually crash into the program load.

**Comment by yuzibo** [comment]
Thanks for help. 

But based on the [diff](https://github.com/sphde/sphde/compare/f3be0d8d14b939fc607a2db3688016291f408e65..d534de21e2e6d6964a83621bc7e95c5d113c7e8f), I am still get unexpected results:

On riscv64, it seems it failed on  sassim test:
```
make[3]: Entering directory '/home/vimer/build/sphde/sphde/src'
PASS: sphgtod_t
PASS: sphgettime_t
PASS: sasatom_t
PASS: bitvec_t
Terminated
make[3]: *** [Makefile:1913: sassim_t.log] Error 124
make[3]: Leaving directory '/home/vimer/build/sphde/sphde/src'
make[2]: *** [Makefile:1867: check-TESTS] Error 2
make[2]: Leaving directory '/home/vimer/build/sphde/sphde/src'
make[1]: *** [Makefile:2102: check-am] Error 2
make[1]: Leaving directory '/home/vimer/build/sphde/sphde/src'
make: *** [Makefile:544: check-recursive] Error 1
```
But there is no ` sassim_t.log`.  trying with 
```
define __SAS_BASE_ADDRESS 0x3000000000L
define RegionSize 0x0800000000L
```

The error is the same.

On riscv32, there are total 13 test cases failed:
```
FAIL: sphdirectpcqueue_ttt
============================================================================
Testsuite summary for libsphde 1.4.0
============================================================================
# TOTAL: 23
# PASS:  8
# SKIP:  2
# XFAIL: 0
# FAIL:  13
# XPASS: 0
# ERROR: 0


FAIL: sassim_t
FAIL: sascompoundheap_t
FAIL: sasseg_t
FAIL: sphlflogger_tt
FAIL: sphcontext_t
FAIL: sasindex_t
FAIL: sasindex_tt
FAIL: sasstringbtree_t
FAIL: sasstringbtree_tt
FAIL: sphsinglepcqueue_tt
FAIL sphdirectpcqueue_ttt 

```

Some kernel crash log is on [here](https://gist.github.com/yuzibo/c1429d8266da891712c1a3657f849c71)

**Comment by munroesj52** [comment]
<<On riscv64, it seems it failed on sassim test:

The first thing sassim_t does is call SASJoinRegion () and print:
```
  printf ("__SAS_BASE_ADDRESS=%lx\n", __SAS_BASE_ADDRESS);
  printf ("RegionSize        =%lx\n", RegionSize);
  printf ("SegmentSize       =%lx\n", SegmentSize);
  printf ("__SAS_TEMP_ADDRESS=%lx\n", __SAS_TEMP_ADDRESS);
  printf ("__SAS_TEMP_FREE   =%lx\n", __SAS_TEMP_FREE);
```
If you are not seeing that, SASJoinRegion () is failing which implies file  SASSTOREPATH  is not rw or SAS_BASE is not mmap-able. 

Edit sassim.c and uncomment #define __SASDebugPrint__, rebuild and see if you can get more information.

**Comment by munroesj52** [comment]
Sorry for getting lost in the weed with this, but there are some other changes needed.

Add risk-v for config.sub
Add __riskv && (__riscv_xlen == 64) to conditional 64-bit struct in sasanchr.h at line 27?
Add __riskv && (__riscv_xlen == 64) to conditional 64-bit struct in sasssim.c at line 127?

**Comment by munroesj52** [comment]
Apologies, I forget some details that affect each new platform. Please look
at the last comment.

**Comment by yuzibo** [comment]
> ASDebugPrint

Good news on riscv64

```bash
============================================================================
Testsuite summary for libsphde 1.4.0
============================================================================
# TOTAL: 23
# PASS:  23
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================

```

In fact, I should test the code on another riscv64 hardware earlier. It pass all tests on new riscv64 hardware without any modifying base on d534de21e.

Will look at riscv32 again.

Many thanks.

**Comment by munroesj52** [comment]
Excellent!

SPHDE is a bit of a nightmare for cross platform. When I wrote the early code compilers did not have nice #define's for ENDIAN and int/pointer size.

Also Biarch (32/64) platforms where/are inconsistent with the sizeof() int/long/long long.  No end of trouble with that.

I fear the the implicit assumption is:
- 32-bit: sizeof(long) == sizeof(pointer).
- 64-bit: sizeof(long) == sizeof(long long) == sizeof(pointer).

**Comment by munroesj52** [comment]
Silence. Must be busy,. I merge this pull request as it works for risk-v 64-bit. 

I will try to fix issue  #72 for the general case and so make SPHDE less fragile.

**Comment by munroesj52** [comment]
Merging riskv 64-bit support. Will check back on riskv 32-bit later.

**Comment by yuzibo** [comment]
Sorry for the so late reply!

yeah, I always failed to test it on riscv32 then something other broke my plan.

Thanks for merging this to support riscv64 first. This will help me to solidify the code base and then to test it on rv32 
continuous.

**Comment by munroesj52** [comment]
Please review and if you can test on your platform.

https://github.com/sphde/sphde/pull/75

**Comment by yuzibo** [comment]
> Please review and if you can test on your platform. #75
Thanks for working on this. 

hmm, but unfortunately, but there is one issue that needs to be addressed for riscv64 before testing it.

First, I `git clone` the newest codebase from this repo and checkout pull/75 locally to verify it on riscv64 hardware. but it give me the error:
```
PASS: sphsinglepcqueue_t
PASS: sphsinglepcqueue_tt
Terminated
make[3]: *** [Makefile:2032: sphsinglepcqueue_ttt.log] Error 124
make[3]: Leaving directory '/home/rv/build/sphde-clean/sphde/src'
make[2]: *** [Makefile:1867: check-TESTS] Error 2
make[2]: Leaving directory '/home/rv/build/sphde-clean/sphde/src'
make[1]: *** [Makefile:2102: check-am] Error 2
make[1]: Leaving directory '/home/rv/build/sphde-clean/sphde/src'
...
```

I suspect this due to #75, but I tried it without the PR, the error was still there. 

It seems the `sphsinglepcqueue_ttt` was crashed when executed, but if I execute it alone, this is okay:

```
rv@debci-01:~/build/sphde-clean/sphde/src$ sudo ./sphsinglepcqueue_ttt
[sudo] password for rv:
SAS Joined with 4 processors

SPHLFpcqueueCreate (4096) success
SPHSinglePCQueueFreeSpace() = 3712 units=29

SPHLFpcqueueCreate (4096) success

SPHLFpcqueueCreate (4096) success
start 1 test_thread_Producer | consumer (0x1000103000,29)
...
startt=259702072481795, endt=259746339603491, deltat=44267121696, freqt=1000000000
test_thread_Producer | transfercopy | consumer wait 29000000 ave= 1526.45ns rate=  655113.7/s
end 10   test_thread_Producer | transfercopy | consumer wait (0x1000103000,29) = 0
Producer wait/post 273672/273673 consumer wait/post 115174/115175
Producer2 wait/post 195824/195825 consumer2 wait/post 205217/205218
Producer3 wait/post 104159/104160 consumer3 wait/post 273036/273037
SAS removed
```

Any ideas to debug it? TIA.

**Comment by munroesj52** [comment]
make[3]: *** [Makefile:2032: sphsinglepcqueue_ttt.log] Error 124

This seems to be a file space/access/permistions issue.

"SAS removed" printf before exit and any file write-buffer and close.

Please verify that you have the file space on your test system. SPHDE
likes to allocate
8+ SegmentSize blocks for unit tests and the log file might just hit the limit.

$df -k
$ls -la SAS*.DAT

**Comment by yuzibo** [comment]
>8+ SegmentSize blocks for unit tests and the log file might just hit the limit.

This hints me, thanks. I tested them at that time with running some others building task.

So now all tests can be passed based on #70 or #75 on riscv64.

**Comment by munroesj52** [comment]
Great! Thanks

**Pull request merged**